### PR TITLE
Users go to old home page when trying to submit for new course #4572

### LINKED ISSUE #4562: StudentFeedbackResultsPageUiTest failing on live server
java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
	at teammates.test.pageobjects.AppPage.createNewPage(AppPage.java:1128)
	at teammates.test.pageobjects.AppPage.getNewPageInstance(AppPage.java:160)
	at teammates.test.cases.ui.browsertests.StudentFeedbackResultsPageUiTest.loginToStudentFeedbackResultsPage(StudentFeedbackResultsPageUiTest.java:141)
	at teammates.test.cases.ui.browsertests.StudentFeedbackResultsPageUiTest.testAll(StudentFeedbackResultsPageUiTest.java:110)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:84)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:714)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:901)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1231)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:127)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:111)
	at org.testng.TestRunner.privateRun(TestRunner.java:767)
	at org.testng.TestRunner.run(TestRunner.java:617)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:334)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:329)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:291)
	at org.testng.SuiteRunner.run(SuiteRunner.java:240)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1224)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1149)
	at org.testng.TestNG.run(TestNG.java:1057)
	at org.testng.remote.RemoteTestNG.run(RemoteTestNG.java:111)
	at org.testng.remote.RemoteTestNG.initAndRun(RemoteTestNG.java:204)
	at org.testng.remote.RemoteTestNG.main(RemoteTestNG.java:175)
Caused by: java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:57)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:526)
	at teammates.test.pageobjects.AppPage.createNewPage(AppPage.java:1124)
	... 27 more
Caused by: java.lang.IllegalStateException: Not in the correct page!
	at teammates.test.pageobjects.AppPage.<init>(AppPage.java:142)
	at teammates.test.pageobjects.StudentFeedbackResultsPage.<init>(StudentFeedbackResultsPage.java:9)
	... 32 more

### LINKED ISSUE #4572: Users go to old home page when trying to submit for new course
I have several cases of this.
A student receives a link from a new course.
After clicking on the link, students ends up in his/her home page which shows his previous courses but not the new one.
Some reported seeing the 'expired login' message in the home page. So students get redirected to home page because TEAMMATES detects an expired login.

### PR DESCRIPTION
**Pull request by wkurniawan07**
Fixes #4572 and fixes #4562 (the test failure in the latter has the same cause).
This issue also highlights our alarming failure in handling next URL. Just doing what is suggested [here](https://github.com/TEAMMATES/repo/issues/4572#issuecomment-157297754) is insufficient.
1. `nextUrl` can be anything from the pages that do not require Google login (they are defined in `Const.SystemParams`) for this to happen.
2. Instead of just status message, it is better to change the content of the box entirely because the intention of confirming the Google ID is not exactly the same:
image
3. True enough, encoding the URL doesn't work because behind the screen Google does its own encoding and decoding. One (and maybe the only) workaround is replacing dangerous characters (i.e `&`, `+`, `#`) with placeholders and revert them back afterwards.

Tested and passed in both dev and staging, but will need more time to design test case(s) for this one. Feel free to review the change in functional codes, though.

### CODE REVIEW

File: src/main/java/teammates/ui/controller/StudentCourseJoinAction.java [hunk 1]
```diff
@@ -34,8 +35,13 @@ public ActionResult execute() throws EntityDoesNotExistException {
         
         String confirmUrl = Const.ActionURIs.STUDENT_COURSE_JOIN_AUTHENTICATED 
                 + "?" + Const.ParamsNames.REGKEY + "=" + regkey 
-                + "&" + Const.ParamsNames.NEXT_URL + "=" + nextUrl;
-        data = new StudentCourseJoinConfirmationPageData(account, student, confirmUrl, Logic.getLogoutUrl(confirmUrl));
+                + "&" + Const.ParamsNames.NEXT_URL + "=" + Sanitizer.sanitizeForNextUrl(nextUrl);
+        String nextUrlType = nextUrl.replaceFirst("^(/page/[A-Za-z]+)\\?.*$", "$1");
```
Comment (damithc) [inline src/main/java/teammates/ui/controller/StudentCourseJoinAction.java]: It's good if this kind of thing can be unit tested. e.g. extract into a small static method and write a few unit tests against it.

File: src/main/java/teammates/common/util/Sanitizer.java [hunk 2]
```diff
@@ -131,6 +131,15 @@ public static String sanitizeForUri(String uri) {
...
     
+    public static String sanitizeForNextUrl(String url) {
+        return url.replace("&", "${amp}").replace("%2B", "${plus}").replace("%23", "${hash}");
+    }
+    
+    public static String desanitizeFromNextUrl(String url) {
+        return url.replace("${amp}", "&").replace("${plus}", "%2B").replace("${hash}", "%23")
+                  .replace(" ", "+");
+    }
+
```
Comment (damithc) [inline src/main/java/teammates/common/util/Sanitizer.java]: Public methods. Add header comments?

File: src/main/java/teammates/ui/controller/StudentCourseJoinAuthenticatedAction.java [hunk 3]
```diff
@@ -24,8 +25,16 @@
     @Override
     protected ActionResult execute() throws EntityDoesNotExistException {
         Assumption.assertNotNull(regkey);
```
Comment (damithc) [inline src/main/java/teammates/ui/controller/StudentCourseJoinAuthenticatedAction.java]: If this parameter can contain more than the key, then may be we should not call it regkey?

**Comment by damithc** [comment]
Looks like this part of the code is not covered by tests. Add some?

**Comment by damithc** [comment]
Sorry, didn't notice that it is still ongoing.

**Comment by wkurniawan07** [comment]
True, it is not covered by tests. I'm still figuring out how best to cover these, i.e via action or via UI or both or some other way.

File: src/main/java/teammates/ui/controller/StudentCourseJoinAuthenticatedAction.java (hunk 3 above)
Comment (wkurniawan07) [inline src/main/java/teammates/ui/controller/StudentCourseJoinAuthenticatedAction.java]: Will be hard to do that, because the `regkey` param comes from the base `Action` class. Also, really it shouldn't be the case that regkey contains something else, if not for this `${amp}` workaround. I think the best I can do is adding some comments there.

**Comment by wkurniawan07** [comment]
I've made a shocking discovery: there are two UI tests covering it, one of them is failing at all times and the other one is failing unless run alone, both in staging. I have every reason to believe that they will behave the same in live. Are `StudentFeedbackResultsPageUiTest` and `StudentFeedbackSubmitPageUiTest` passing all these while?

**Comment by damithc** [comment]
StudentFeedbackResultsPageUiTest is currently an open issue. The other is not, probably because it passes when run alone.

**Comment by wkurniawan07** [comment]
:+1: just as I suspected. This is good because both issues have the same fix, but might need some time to implement and test (need to ensure passing in both dev and staging).

File: src/main/java/teammates/ui/controller/StudentCourseJoinAuthenticatedAction.java (hunk 3 above)
Comment (thyageshm) [inline src/main/java/teammates/ui/controller/StudentCourseJoinAuthenticatedAction.java]: Hey, we could consider abstracting this away in Action.java. In the init() function we can check if there are ${amp}s in the complete url. If yes, break it up before coming to the individual Action level. 

Based on what you said, this problem could arise in other places and a common decoding placed at Action.java would help.

Comment (wkurniawan07) [inline src/main/java/teammates/ui/controller/StudentCourseJoinAuthenticatedAction.java]: Not necessary. You can see that the change affected only `StudentCourseJoinAction` in which the next URL is very specific, `Const.ActionURIs.STUDENT_COURSE_JOIN_AUTHENTICATED`.
Until next URLs everywhere else is handled the way it is handled here, only this class needs a special treatment.

**Comment by damithc** [comment]
This issue needs a quick fix. So parts that are not urgent can be done as a separate issue.

**Comment by wkurniawan07** [comment]
I found more and more worrying things as I try to fix this. Turns out there's a trade-off between testing redirection and allowing app URL defined in `build.properties` to be free:
```
# This is the app url used in emails sent out by app.
# For dev server, any URL would do. 
```
This is wrong. In redirection cases due to user being unregistered, the redirect URL will start with student's registration URL in which the app URL will be taken from `build.properties`, not `test.properties`. There is no way to replace it halfway without our testing code polluting the production code. This has never been covered before.
If we require app URL to be a working URL (i.e same as the one defined in `test.properties`), then we cannot use God mode.

**Comment by wkurniawan07** [comment]
It should be possible to test via action test, but I guess that's the best we can do.

**Comment by wkurniawan07** [comment]
Extended the UI test case for `StudentFeedbackResultsPageUiTest` (because this is the shortest) to cover the new display, i.e this:
image
and also what will happen after registering using logged in Google account or using another account after logging out (i.e covering both options). That's three new HTML files.
As mentioned before, the redirection cannot be tested via UI. Still trying to test it via action.

**Comment by damithc** [comment]
We can deploy this first (if manual tests are working) and add the tests later.

**Comment by wkurniawan07** [comment]
The redirections work on both dev and staging, and so is the UI test mentioned above, only missing the action test to cover the redirection path (if it can be covered). So I guess it's ready for review.

**Comment by wkurniawan07** [comment]
Good news: managed to cover redirects via UI. Turns out the problem lies in `StudentAttributes` itself, where `getRegistrationUrl()` returns absolute path (it should be relative path). In other words, this case is now fully covered with test.
Note that I add the test in `StudentFeedbackResultsPageUiTest` for the reason I mentioned before (i.e shorter as compared to `StudentFeedbackSubmitPageUiTest` and `StudentFeedbackQuestionSubmitPageUiTest`) but I believe that the behavior of redirects will be the same in the other two.

Now if someone wants to review this...

**Comment by damithc** [comment]
@thyageshm is the best person to review this.

File: src/test/java/teammates/test/cases/ui/browsertests/StudentFeedbackSubmitPageUiTest.java [hunk 4]
```diff
@@ -78,6 +78,9 @@ private void testLinks() {
     }
 
     private void testContent() {
+
+        AppPage.logout(browser);
```
Comment (thyageshm) [inline src/test/java/teammates/test/cases/ui/browsertests/StudentFeedbackSubmitPageUiTest.java]: I think we should put this line after **______TS("unreg student");** as this is needed for that test specifically. If someone moves that test, this needs to move as well

File: src/main/java/teammates/common/datatransfer/StudentAttributes.java [hunk 5]
```diff
@@ -130,7 +129,7 @@ public boolean isRegistered() {
     }
 
     public String getRegistrationUrl() {
-        return new Url(Config.APP_URL + Const.ActionURIs.STUDENT_COURSE_JOIN_NEW)
+        return new Url(Const.ActionURIs.STUDENT_COURSE_JOIN_NEW)
```
Comment (thyageshm) [inline src/main/java/teammates/common/datatransfer/StudentAttributes.java]: Hey could you tell me the reason why we changed this to relative path?

Comment (wkurniawan07) [inline src/main/java/teammates/common/datatransfer/StudentAttributes.java]: Consider going to a link that needs redirect, in which the redirection path takes you to the registration URL with the original URL as the next URL. Let's say the link is `/page/xyz`. The registration URL previously will include the app URL, which makes it an absolute path, e.g `http://app.url/registration?next=/page/xyz`.
1. To make it work the app URL needs to be correct. This is against what is written on `build.properties` (`For dev server, any URL would do.`).
2. Absolute path is totally unnecessary unless for presentation purpose (e.g sending the URLs in emails). See the URL just right after that, the `getPublicProfilePictureUrl()`? That one is relative, it works, and it is clearly the better choice.

Comment (wkurniawan07) [inline src/main/java/teammates/common/datatransfer/StudentAttributes.java]: For the very same reason, take a look at the changes I made in `PageData.java`: I believe that they're unnecessary, but I want to keep the diff minimum for now (otherwise I'm changing a few other HTML files; or should I go down this path?).

Comment (damithc) [inline src/main/java/teammates/common/datatransfer/StudentAttributes.java]: I suggest either rename the method to add 'relative' and/or add a comment to explain its behavior.

Comment (wkurniawan07) [inline src/main/java/teammates/common/datatransfer/StudentAttributes.java]: The other URL method `getPublicProfilePictureUrl()` returns relative URL without any indication whatsoever. Come to think of it, it was very odd before: one method returned absolute URL and the other returned relative. Good that it's resolved now.
It should be either nothing is changed or both methods are renamed to `getRelative*Url()`.

Comment (damithc) [inline src/main/java/teammates/common/datatransfer/StudentAttributes.java]: Minimally, add comments to both. I prefer to change the names too (a name is harder to miss than a comment), but it might make them too long. What do you think?

Comment (wkurniawan07) [inline src/main/java/teammates/common/datatransfer/StudentAttributes.java]: I still think it's unnecessary. If anything, it's because system-wide all URLs generated are relative, except for presentation purpose (in emails, in admin search, in admin home page right after account creation). No point in explaining what is commonly practiced.
Having said that, I'm now inclined to use relative URLs also in `PageData.java`. Will do that in a jiffy.

Comment (damithc) [inline src/main/java/teammates/common/datatransfer/StudentAttributes.java]: Sure, we can make it the convention. Note that some devs may not know what the 'system-wide convention' is, especially contributors. Let's hope the respective reviewer will remember to enforce the convention when a contributor breaks it.

Comment (thyageshm) [inline src/main/java/teammates/common/datatransfer/StudentAttributes.java]: I think this change need not be part of this fix. We can create another one to make these changes. Being a very high priority issue, I think we should merge this pr soon.

For the question of whether we should make relative URLs the norm in the system, I am ok with it. In fact, if it works better than full URLs, we should move over. But since there is a real use-case for the full url, we should create a ```Full*Url()``` and ```Relative*Url()``` for clarity and minimise repetitive code. Also, this allows for unit testing the full URL as compared to testing it in ActionTests or browser tests

Comment (wkurniawan07) [inline src/main/java/teammates/common/datatransfer/StudentAttributes.java]: 1. I couldn't test the redirects before. I need either the URL to be relative or the app URL in `build.properties` to be enforced to be a working URL. The latter will render God mode unavailable (because then `app.url` and `test.url` must always be the same) and is also against what is written (`For dev server, any URL would do.`), so it's quite clear that the former (i.e this fix) is the way to go.
2. Relative URL is _already_ the norm in the system. There are only very few exceptions: this case (fixed), the feedback session copy modal (should be fixed), the three I mentioned above (emails, admin search, admin home page right after account creation; these three use absolute URLs correctly), and one in `AdminEmailImageUploadAction` (haven't investigated whether this can be relative, and certainly not here if I'm going to). Everything else other than those are relative URLs.

In other words, yes I can defer the `PageData.java` fix to another issue, but everything else needs to be in this fix.

Comment (thyageshm) [inline src/main/java/teammates/common/datatransfer/StudentAttributes.java]: I understand that the changes are necessary. But we can do something like what was done in #4546. Add the test cases in the new issue and we can merge this one soon. If you feel all the above can be done quickly, I guess there is no harm in it

Comment (wkurniawan07) [inline src/main/java/teammates/common/datatransfer/StudentAttributes.java]: It's already done. The extensions in `StudentFeedbackResultsPageUiTest` are testing the redirects.

**Comment by wkurniawan07** [comment]
Rebased for commit organization and made the change suggested.
I will also add that `studentFeedbackResultsPageNewlyRegistered.html` is a carbon copy of `unregisteredStudentFeedbackResultsPageOpen.html` except for the status messages (I have verified this via Git), and this must be the case because they both display results of the same session (Open Session) for the same student (DropOut).

File: src/main/java/teammates/ui/controller/Action.java [hunk 6]
```diff
@@ -238,10 +238,8 @@ protected AccountAttributes authenticateAndGetNominalUser(UserType loggedInUserT
                 
                 throw new UnauthorizedAccessException("Unregistered user for a page that needs one");
             } else if (isPageNotCourseJoinRelated() && doesRegkeyBelongToUnregisteredStudent() && isUserLoggedIn) {
-                Url redirectUrl = new Url(Const.ViewURIs.LOGOUT)
-                                      .withParam(Const.ParamsNames.NEXT_URL, requestUrl)
-                                      .withUserId(StringHelper.encrypt(account.googleId)); 
-                
+                Url redirectUrl = new Url(student.getRegistrationUrl())
+                                      .withParam(Const.ParamsNames.NEXT_URL, requestUrl);
```
Comment (thyageshm) [inline src/main/java/teammates/ui/controller/Action.java]: hey I think you have to add the UserId parameter so that the admin will be able to masquerade as the student using this redirect

File: src/main/java/teammates/ui/controller/StudentCourseJoinAction.java (hunk 1 above)
Comment (thyageshm) [inline src/main/java/teammates/ui/controller/StudentCourseJoinAction.java]: @wkurniawan07 do you think this can be done?

File: src/main/java/teammates/ui/controller/StudentCourseJoinAuthenticatedAction.java (hunk 3 above)
Comment (thyageshm) [inline src/main/java/teammates/ui/controller/StudentCourseJoinAuthenticatedAction.java]: Ok!

File: src/main/java/teammates/common/datatransfer/StudentAttributes.java (hunk 5 above)
Comment (thyageshm) [inline src/main/java/teammates/common/datatransfer/StudentAttributes.java]: nice :-)

**Comment by thyageshm** [comment]
just a few minor things, other than that it is good to go

File: src/main/java/teammates/ui/controller/StudentCourseJoinAction.java (hunk 1 above)
Comment (wkurniawan07) [inline src/main/java/teammates/ui/controller/StudentCourseJoinAction.java]: Not that it can't be done, but... I don't really see the point why. The method is hardly that complicated to warrant its own test. Also I don't know where I should extract the method to. Doesn't seem to fit to any of the util class.

File: src/main/java/teammates/ui/controller/Action.java (hunk 6 above)
Comment (wkurniawan07) [inline src/main/java/teammates/ui/controller/Action.java]: But this code block is inside `!isMasqueradeModeRequested(...)` which means that it won't be reached if it's the case of masquerading.

File: src/main/java/teammates/ui/controller/StudentCourseJoinAction.java (hunk 1 above)
Comment (damithc) [inline src/main/java/teammates/ui/controller/StudentCourseJoinAction.java]: It can be a static method in this class itself?
Whenever we use regex (not just in this case), I'd like to see the behaviour specified in terms of comments and tests because we can't expect all devs to know regex. Extracting it to small static method gives us a chance to do both. A regex has a lot of hidden logic in it that is not even captured by coverage tools. A unit test targeting only this line will give us an opportunity to write a whole lot of cheap unit tests that illustrating the behavior using negative and positive test cases.

Comment (wkurniawan07) [inline src/main/java/teammates/ui/controller/StudentCourseJoinAction.java]: Added a few tests for the said method. I find it quite strange that an Action test tests something other than `executeAndPostProcess`, though.

File: src/main/java/teammates/ui/controller/StudentCourseJoinAction.java [hunk 7]
```diff
@@ -62,4 +68,21 @@ protected ActionResult createRedirectToAuthenticatedJoinPage(String nextUrl) {
...
+     * <code>/page/xyz?param1=value1&amp;param2=value2</code> is <code>/page/xyz</code>.
+     * The page type is assumed to be in the form of /page/ followed by alphabets
+     * (case-insensitive) only, as per the design of {@link Const.ActionURIs}.
+     */
+    public static String getPageTypeOfUrl(String url) {
+        /* 
+         * Regex meaning: from the beginning of the string, tries to match /page/
+         * followed by one or more case-insensitive alphabets, followed by ? and
+         * any amount of any character until the end of the string.
+         * Returns everything before ? if matches.
```
Comment (damithc) [inline src/main/java/teammates/ui/controller/StudentCourseJoinAction.java]: returns original string if no match?

Comment (wkurniawan07) [inline src/main/java/teammates/ui/controller/StudentCourseJoinAction.java]: That's right. I should mention that also.

File: src/main/java/teammates/ui/controller/StudentCourseJoinAction.java [hunk 8]
```diff
@@ -34,8 +35,13 @@ public ActionResult execute() throws EntityDoesNotExistException {
         
         String confirmUrl = Const.ActionURIs.STUDENT_COURSE_JOIN_AUTHENTICATED 
                 + "?" + Const.ParamsNames.REGKEY + "=" + regkey 
-                + "&" + Const.ParamsNames.NEXT_URL + "=" + nextUrl;
-        data = new StudentCourseJoinConfirmationPageData(account, student, confirmUrl, Logic.getLogoutUrl(confirmUrl));
+                + "&" + Const.ParamsNames.NEXT_URL + "=" + Sanitizer.sanitizeForNextUrl(nextUrl);
+        String nextUrlType = getPageTypeOfUrl(nextUrl);
+        boolean isRedirectResult = Const.SystemParams.PAGES_ACCESSIBLE_WITHOUT_GOOGLE_LOGIN.contains(nextUrlType);
```
Comment (damithc) [inline src/main/java/teammates/ui/controller/StudentCourseJoinAction.java]: This sounds like a shortcut that might work now but could break later. It says, if a page is accessible without Google login, then it is a redirect result. Are the two things going to be equivalent all the time? or should we have another set called `PAGES_RESULTING_FROM_REDIRECTS` or something like that? Or is there a more intuitive way to check if this is a redirect result?

File: src/test/java/teammates/test/cases/ui/StudentCourseJoinActionTest.java [hunk 9]
```diff
@@ -109,6 +111,22 @@ public void testExecuteAndPostProcess() throws Exception {
...
+        assertEquals("/page/somePage", getPageTypeOfUrl("/page/somePage?key=abcdef&next=/page/anotherPage?"));
+        // the starting keyword /page/ must be strictly followed
+        assertEquals("/pag/somePage?key=abcdef", getPageTypeOfUrl("/pag/somePage?key=abcdef"));
+        assertEquals("page/somePage?key=abcdef", getPageTypeOfUrl("page/somePage?key=abcdef"));
+        // and must strictly be at the start of the string
+        assertEquals("abc.com/page/somePage?key=abcdef", getPageTypeOfUrl("abc.com/page/somePage?key=abcdef"));
+        // non-alphabet characters are not handled
+        assertEquals("/page/somePage123?key=abcdef", getPageTypeOfUrl("/page/somePage123?key=abcdef"));
+        assertEquals("/page/somePage/somePage?key=abcdef", getPageTypeOfUrl("/page/somePage/somePage?key=abcdef"));
+    }
```
Comment (damithc) [inline src/test/java/teammates/test/cases/ui/StudentCourseJoinActionTest.java]: Nice.

**Comment by damithc** [comment]
Let's try to merge this today.

File: src/main/java/teammates/ui/controller/StudentCourseJoinAction.java (hunk 8 above)
Comment (thyageshm) [inline src/main/java/teammates/ui/controller/StudentCourseJoinAction.java]: It is true. Here is a case where we might want to redirect to the course join page from one of the non-```PAGES_ACCESSIBLE_WITHOUT_GOOGLE_LOGIN``` pages:

We want to promote some of our "registered features" like comments. So we allow instructors to comment on student responses. We then email the student saying that the instructor has some comments on his/her response and he/she can register with TEAMMATES to view it. We can then give them the registration link with the next url as the comments page in the email itself.

File: src/main/java/teammates/ui/controller/Action.java [hunk 10]
```diff
@@ -238,10 +238,9 @@ protected AccountAttributes authenticateAndGetNominalUser(UserType loggedInUserT
                 
                 throw new UnauthorizedAccessException("Unregistered user for a page that needs one");
             } else if (isPageNotCourseJoinRelated() && doesRegkeyBelongToUnregisteredStudent() && isUserLoggedIn) {
-                Url redirectUrl = new Url(Const.ViewURIs.LOGOUT)
-                                      .withParam(Const.ParamsNames.NEXT_URL, requestUrl)
-                                      .withUserId(StringHelper.encrypt(account.googleId)); 
-                
+                Url redirectUrl = new Url(student.getRegistrationUrl())
+                                      .withParam(Const.ParamsNames.STUDENT_NEW_COURSE, student.course)
+                                      .withParam(Const.ParamsNames.NEXT_URL, requestUrl);
```
Comment (wkurniawan07) [inline src/main/java/teammates/ui/controller/Action.java]: Ok, so now here is how it works: We know that if this code block is reached, these conditions must be fulfilled:
1. There is a logged in user
2. The user is unregistered for the new course
3. The user is registered in some other course(s)
4. The user navigates to a URL other than the join URL
In this case we pass the ID of the new course as a parameter. The existence of this parameter will then decide whether the page looks like this (if exists):
image
or this (if doesn't exist):
image
This should cover all cases. Note that the difference is purely cosmetic. On both pages' counterparts, the buttons will direct to the same URL.

Comment (thyageshm) [inline src/main/java/teammates/ui/controller/Action.java]: Hey, I would like to point out that (3) is not always true as the user *CAN* be loggedin without a course (although highly unlikely)

Comment (wkurniawan07) [inline src/main/java/teammates/ui/controller/Action.java]: If you mean the user is a total stranger, then it would be captured by the block just before it (`doesUserNeedRegistration(account)`).
If you mean the user was registered before but his/her courses were all deleted, then yes 3 includes such case as well. But really what are the chances of this happening.

File: src/main/java/teammates/ui/controller/StudentCourseJoinAction.java [hunk 11]
```diff
@@ -34,8 +35,12 @@ public ActionResult execute() throws EntityDoesNotExistException {
...
         String confirmUrl = Const.ActionURIs.STUDENT_COURSE_JOIN_AUTHENTICATED 
                 + "?" + Const.ParamsNames.REGKEY + "=" + regkey 
-                + "&" + Const.ParamsNames.NEXT_URL + "=" + nextUrl;
-        data = new StudentCourseJoinConfirmationPageData(account, student, confirmUrl, Logic.getLogoutUrl(confirmUrl));
+                + "&" + Const.ParamsNames.NEXT_URL + "=" + Sanitizer.sanitizeForNextUrl(nextUrl);
+        String studentNewCourse = getRequestParamValue(Const.ParamsNames.STUDENT_NEW_COURSE);
+        boolean isUserLoggedInAsAnotherStudent = studentNewCourse != null;
+        data = new StudentCourseJoinConfirmationPageData(account, student, confirmUrl,
+                                                         Logic.getLogoutUrl(Sanitizer.sanitizeForNextUrl(confirmUrl)),
+                                                         isUserLoggedInAsAnotherStudent, studentNewCourse);
```
Comment (thyageshm) [inline src/main/java/teammates/ui/controller/StudentCourseJoinAction.java]: two things here:
1. the name ```isUserLoggedInAsAnotherStudent``` is confusing
2. the course that we are sending in as a parameter is the same as ```student.course``` which is obtained from the regkey in the registration link.

I understand that you are trying to fix the problem by adding a new parameter that indicates whether a special action is needed or not. I believe the simplest solution is what you had done earlier, but instead of checking if the next_url is in the ```PAGES_ACCESSIBLE_WITHOUT_GOOGLE_LOGIN``` list, check if the next_url is NOT IN ```PAGES_ACCESSIBLE_WITHOUT_REGISTRATION```. This is future-proof and technically this is what we should be targeting. 

Just my thoughts...

File: src/main/java/teammates/ui/controller/Action.java (hunk 10 above)
Comment (thyageshm) [inline src/main/java/teammates/ui/controller/Action.java]: You are right, it is very, very unlikely. But more importantly I guess it is not an issue in our case, whether or not the user is registered in another course is irrelevant here

File: src/main/java/teammates/ui/controller/StudentCourseJoinAction.java (hunk 11 above)
Comment (wkurniawan07) [inline src/main/java/teammates/ui/controller/StudentCourseJoinAction.java]: Makes sense. Let's see if it addresses the issue reliably.

Comment (wkurniawan07) [inline src/main/java/teammates/ui/controller/StudentCourseJoinAction.java]: It works, and looks like it is the better alternative after all. Should be good to go now, I guess?

File: src/main/webapp/jsp/studentCourseJoinConfirmation.jsp [hunk 12]
```diff
@@ -15,18 +15,47 @@
...
+	            <c:if test="${data.redirectResult}">
+	                You have been redirected to this page because you navigated to a link for the course <strong>${data.courseId}</strong>, which you have not been registered in.
+	            </c:if>
 	            <br>
+	            <c:choose>
+	                <c:when test="${data.redirectResult}">
+	                    If you wish to register as <strong>${data.account.googleId}</strong>, please confirm below to complete your registration.
+	                    <br>If you wish to register with another Google ID or do not wish to register for this course, please
+	                    <a href="${data.logoutUrl}">log out</a>
+	                    , log in with your desired Google ID if necessary, and navigate to the link sent to you via email again.
```
Comment (thyageshm) [inline src/main/webapp/jsp/studentCourseJoinConfirmation.jsp]: Now that we are using this message whenever the next_url is a page that requires registration, we might have to update the message here for future-proofing. Especially the part about "navigating to the link that came via email" needs to be updated as now the user may not be allowed to access the next_url without registering at all...

**Comment by thyageshm** [comment]
Great work! Just one last gap to consider for future proofing (not really affecting this issue)

File: src/main/webapp/jsp/studentCourseJoinConfirmation.jsp (hunk 12 above)
Comment (wkurniawan07) [inline src/main/webapp/jsp/studentCourseJoinConfirmation.jsp]: Now comes the distinction between can access without registering and can access without login. An easy way out is to make the words "or do not wish to register for this course" and "if necessary" spelled out only if the next URL is accessible without login.
And I'm thinking to remove "sent to you via email" because for all we know the link may come from somewhere else.
Sounds good?

**Comment by wkurniawan07** [comment]
This is what it will look like if the next URL requires login:
image
Only slight difference but I believe it's enough to tell that not registering/not logging in is not an option.

**Comment by thyageshm** [comment]
I am ok with it :-) @damithc any comments?

If nothing, the pr is good to go!

**Comment by damithc** [comment]
:+1:

**Comment by wkurniawan07** [comment]
Came back from lunch to this.
image

**Comment by thyageshm** [comment]
:+1:
